### PR TITLE
Optimize about page

### PR DESCRIPTION
**Pull request by ethanliu24**
## Proposed Changes

_(Describe your changes here. Also describe the motivation for your changes: what problem do they solve, or how do they improve the application or codebase? If this pull request fixes an open issue, [use a keyword to link this pull request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)_

Each time we go to the about page, a markdown file is compiled to HTML and then served from the backend. This can be inefficient, so this PR aims to improve the load efficiency by letting Webpack to precompile the markdown and serve that rather than compiling everything again.

Also in the about page, when we click on links to learn about privacy and licensing, it is directed to an invalid path. This PR also fixes this issue.
...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| 🎨 _User interface change_ (change to user interface; provide screenshots)              |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |X         |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [ ] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [ ] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [ ] If this is my first contribution, I have added myself to the list of contributors.
- [x] I have updated the project Changelog (this is required for all changes).

After opening your pull request:

- [x] I have verified that the CircleCI checks have passed.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

**Comment by ethanliu24** [comment]
Hey Professor Liu, in Slack I mentioned that I couldn't get `remark-html` to parse the markdown file properly: it's parsed as a serialized markdown string rather than HTML. This draft is what I have so far with this solution.

In `webpack.common.js`, the commented part is trying to use remark-html and remark-loader (compiled to markdown string, following https://webpack.js.org/loaders/remark-loader/#markdown-to-html), and the code below it is using markdown-loader (compiled to HTML string, the desired output). I have attacked screen shots of what both looks like. 

This is compiled into `./public` and I injected it in the backend. I am also not too sure about how I could retrieve the serialized string from it; I have a few ideas, though I'm not sure if they are good or not:

1. In Haskell, I can read the compiled js script and use regex to retrieve the string.

2. Instead of processing the markdown file directly in Webpack, I could write a new script and using `remark-html` directly (following https://github.com/remarkjs/remark-html?tab=readme-ov-file#use) to compile it, and insert it in the DOM. I don't think this is a good solution though, as we are recompiling the markdown over and over again, just in the front end this time.

3. Almost a combination of the previous two ideas, I could write a new script and during compile time, the serialized string could be written to a new file in `./public` (maybe a txt or html file) and the backend can read and insert it in DOM so we don't have to recompile repeatedly like idea #2?

Screenshot 2024-10-26 at 5 42 40 PM
Screenshot 2024-10-26 at 5 44 09 PM

**Review by david-yz-liu — COMMENTED** [review]
Hi @ethanliu24, thanks for making this PR and your detailed report. I played around with `remark-html` myself, and replicated the issue you reported (no conversion to HTML), and was unable to figure out how to fix it myself! Strange. So, I'm happy to stick with the alternate library you found, and great research in finding it.

Now, for your question about how to incorporate the HTML: the generated file is a Javascript file that has access to the raw HTML. Right now you just have `README.md` as the entrypoint file, but you can instead create a separate `about.js` file with contents like

```javascript
# about.js
import aboutContents from '../path/to/README.md'

# aboutContents is the HTML string
console.log(aboutContents)
```

and then from there dynamically insert the contents into the webpage. It's not as smooth as if the page were a static HTML file being served, but it should be pretty fast.

**Comment by ethanliu24** [comment]
Hey Professor Liu I've made changes to your comment. The raw HTML is now dynamically inserted into the DOM via scripts in `js/components/about` (I'm not sure where to put these scripts, since they are not react components). Is this the way of inserting HTML that you were looking for?

Also as I was fixing the `privacy` route, I noticed that it can't be triggered from anywhere. When we click on the privacy link in the About page, it directs to route `PRIVACY.md` because that's how it was directed in the markdown file. I think the easiest way to fix that is either by adding a `PRIVACY.md` route or redirect to `privacy` when we are on `PRIVACY.md`. Currently I have the former, but which one do you prefer, or if there's a better way to do it?

Last, I noticed that the `LICENSE` link is also not loaded with the same reasoning as privacy. Should I add a new license route and fix that also?

File: app/Routes.hs [hunk 1]
```diff
@@ -48,8 +47,9 @@ strictRoutes aboutContents privacyContents = [
     ("post", retrievePost),
     ("post-progress", postResponse),
     ("draw", drawResponse),
-    ("about", aboutResponse aboutContents),
-    ("privacy", privacyResponse privacyContents),
+    ("about", aboutResponse),
+    ("PRIVACY.md", privacyResponse),  -- check for this route in strictMatchDir and redirect to /privacy?
```
Comment (david-yz-liu) [inline app/Routes.hs:51]: Okay, I looked into this some more. I don't actually think it's worth generating the privacy policy as an HTML file, and instead I'd like to modify the link to refer directly to https://github.com/Courseography/courseography/blob/master/PRIVACY.md, and similarly for the LICENSE. Please make those changes in the README.md file, and then *remove* all privacy-policy-related routes/functions/files in this PR.

File: app/Server.hs [hunk 2]
```diff
@@ -9,10 +9,9 @@ responses.
 module Server
     (runServer) where
 
-import Config (markdownPath, serverConf)
+import Config (serverConf)
```
Comment (david-yz-liu) [inline app/Server.hs:12]: This is great; you can check whether `markdownPath` is used anywhere else, and (assuming not), remove it entirely from `Config` and the configuration YAML file.

File: webpack.common.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 const path = require("path")
 const MiniCssExtractPlugin = require("mini-css-extract-plugin")
 const webpack = require("webpack")
+// const RemarkHTML = import("remark-html")
```
Comment (david-yz-liu) [inline webpack.common.js:4]: remove this line

**Review by david-yz-liu — COMMENTED** [review]
@ethanliu24 great work! I left some inline comments, and also I think you forgot to commit the changes in the `.yarn` folder, which is where the actual JS dependencies are stored.

**Comment by ethanliu24** [comment]
It seems like when I do a `yarn install` to install the dependencies I added, quite a few existing dependencies have been deleted. Is this an expected behaviour?

**Review by david-yz-liu — COMMENTED** [review]
@ethanliu24 it looks like you accidentally upgraded various NPM dependencies, either earlier in the semester or when adding the new dependencies. You can fix this by doing the following:

1. Revert *all* changes to `.yarn`, `package.json`, and `yarn.lock`.
2. Run `yarn install` again, and verify with a `git status` that *nothing is changed* (because at this point your `package.json` should be the same as what's on master).
3. Add the two packages to `package.json`.
4. Run `yarn install` again.

You should then see (with a `git status` that very few packages in `.yarn` have been updated, and that the `yarn.lock` file has just the new dependencies added (plus any of their transitive dependencies).

File: js/components/about/about.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import aboutContent from "../../../README.md"
+
+document.querySelector("#aboutDiv").innerHTML = aboutContent
```
Comment (david-yz-liu) [inline js/components/about/about.js:3]: Two comments:

1. It's more idiomatic to use `document.getElementById` here
2. Move this code into a [`DOMContentLoaded` event callback](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event)

**Review by david-yz-liu — COMMENTED** [review]
@ethanliu24 great work! Everything is looking very good, I just left two quick comments.

I agree the javascript file placement isn't ideal, but I think it's fine for now (appealing to consistency with other pages).

**Comment by david-yz-liu** [comment]
@ethanliu24 oh one more thing, please add `public/js/about` to the `.gitignore` (see where the other built files are being listed in that file)

**Review by david-yz-liu — APPROVED** [review]

**Pull request merged**